content/webapp/utils/digital-guides.ts

```diff
@@ -0,0 +1,100 @@
+ * == Historical note ==
+ *
+ * The original implementation of this was more aggressive: in particular,
+ * it would redirect a user who landed on a guide page, even if they didn't
+ * come from scanning a QR code.
+ *
+ * We changed it after discovering it broke the "Back" button:
+ *
+ *    1.  A user goes to an exhibition guide overview page `/guides/exhibitions/[id]`
+ *    2.  They click to select a particular guide, say `/guides/exhibitions/[id]/audio`.
+ *        This sets a preference cookie telling us the user wants audio guides.
+ *    3.  They click the "Back" button in their browser because they want to see the
+ *        page they were just looking at. This takes them to `/guides/exhibitions/[id]`
+ *        … where we redirect them back to the guide they were just looking at.
+ *
+ * We only create QR codes that link to the audio guides, not the overview page, so
+ * this prevents somebody getting stuck in this sort of redirect loop.
+ *
+ */
+
+export const getGuidesRedirections = (
+  context: GetServerSidePropsContext<ParsedUrlQuery, PreviewData>
+) => {
+  const { req, res, resolvedUrl } = context;
+  const { id: guideId, stopNumber, stopId, type, usingQRCode } = context.query;
+
+  if (!usingQRCode) return;
+
+  const userPreferenceGuideType = getCookie(cookies.exhibitionGuideType, {
+    req,
+    res,
+  });
+
+  const hasValidUserPreference =
+    typeof userPreferenceGuideType === 'string' &&
+    userPreferenceGuideType !== type &&
+    isValidExhibitionGuideType(userPreferenceGuideType);
+
+  // Supporting Jason exhibition
+  // TODO remove when it closes/we adapt the QR codes
+  if (
+    guideId === 'Zdcs4BEAACMA6abC' &&
+    hasValidUserPreference &&
+    type &&
+    typeof stopId === 'string'
+  ) {
+    return {
+      redirect: {
+        permanent: false,
+        // We do a simple replace on the URL so we preserve all other URL information
+        // (e.g. UTM tracking parameters).
+        destination: context.resolvedUrl.replace(
+          `/${type}`,
+          `/${userPreferenceGuideType}`
+        ),
+      },
+    };
+  }
+
+  // New exhibition QR codes URLs should _always_ have the following format:
+  // guides/exhibitions/[exhibitionId]?stopNumber=[stopNumber]
+  const hasValidStopNumber =
+    typeof stopNumber === 'string' && !!Number(stopNumber);
+
+  // The first stop's link is an exception, in that it should link to
+  // the [exhibitionId]/[type] page, and not directly on the stop's page.
+  if (hasValidUserPreference && hasValidStopNumber) {
+    // We do a simple replace on the URL so we preserve all other URL information
+    // (e.g. UTM tracking parameters).
+    return {
+      redirect: {
+        permanent: false,
+        destination: resolvedUrl
+          .replace(
+            '?',
+            `/${userPreferenceGuideType}/${stopNumber === '1' ? '' : stopNumber}?`
+          )
+          .replace(`&stopNumber=${stopNumber}`, '')
+          .replace(`?stopNumber=${stopNumber}`, ''),
```

Comment:
There's surely a better way to do this. Like removing the character before whatever it is 🤔